text/0000-decorator-support.md

```diff
@@ -0,0 +1,142 @@
+forward to native class syntax without a native equivalent to "classic"
+decorators.
+
+[Stage 2 in TC39](https://tc39.github.io/process-document/) signals that the
+committee expects this feature to be included in the language in the future, and
+is working out the details and semantics. The fact that decorators _remain_ in
+stage 2, and have not been rejected, signals that they still expect this to be
+the case. This is also the feedback that the core team received in the January
+TC39 meeting. _Crucially_, all parties were in agreement about the _invocation_
+syntax of decorators:
+
+```js
+class Person {
+  @decorate name;
+}
+```
+
+This is the most stable part of the spec, and the one that average developers
+will come into contact with most. It means that changes to the spec will likely
+affect the _application_ of decorators, which will in turn affect library and
+framework maintainers, but **not** end users in most cases. This places the
+burden of dealing with the spec changes on the Ember.js core team, and addon
+authors who wish to adopt decorators while they are in stage 2.
+
+As such, adopting decorators now should present a minimal amount of risk to
+Ember and its users, and while it is not ideal, it allows us to continue moving
+forward and innovating as a framework.
+
+## Detailed design
+
+### Classic Classes
+
+The first thing to note is that Ember.js will continue to work with classic
+classes and classic class syntax for the forseeable future, as a way for users
+to opt-out if they don't want to take the risk of using native classes. This
+includes:
+
+- Classic Classes
+- Classic Components
+- Mixins
+- Computed Properties
+- Observers
+- Tracked Properties
+- Injections
+- All existing classes defined with classic class syntax
+
+Notably, `GlimmerComponent` will _not_ support classic class syntax, due to its
+constraint of having to support both Ember.js _and_ Glimmer.js, where the
+classic class model is not available. However, creating an API compatible
+classic class version using Ember's component manager API should be possible if
+users want to write Glimmer-like component's with classic syntax.
+
+### Decorator Versioning and Support
+
+Because the decorators spec has changed while remaining in a stage 2, it's no
+longer enough to refer to "stage 2" as a version. Since Babel will be providing
+the transforms, we will use their versioning system. [This is currently in
+development](https://github.com/babel/babel/pull/9360), but looks like it will
+be based on the TC39 meeting that the spec was presented at. For instance, the
+current transforms available in Babel are referred to as `nov-2018`.
+
+Ember will begin by supporting the _latest_ version of the decorators transform
+provided by Babel. Currently this is the `nov-2018` version, but by the time
+decorators are released this may change.
+
+The decorators version will be configurable on a _per-app/engine/addon_ basis.
```

Comment:
We are concerned about this, yes, and we have been talking about ways to mitigate it. The current strategy we have in mind is to allow users to define or build multiple files for each decorator version implemented, and then have a babel transform rewrite the import paths to the version desired in a given addon. Combined with the ongoing work to revamp the build system for packaging and treeshaking, this would allow us to treeshake out any fully unused versions.

This plan is not outlined here however because it may not be possible, and relies on too many unknowns. In the worst case, we may have to continue shipping multiple versions of the transforms for some amount of time. Given that the transforms do not update often, and we are retaining the ability to drop support for versions after LTS releases, this should not amount to much extra weight of code.